File: tests/c++/cgns/test_base_tree.cpp
#include "test_base_tree.hpp"

#include "array/array.hpp"
#include "array/array_numpy_bridge.hpp"
#include "cgns/base.hpp"
#include "cgns/tree.hpp"
#include "cgns/zone.hpp"

#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>

#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

namespace py = pybind11;

namespace {

std::shared_ptr<Zone> makeCartesianZone(
    const std::string& name,
    const std::vector<ssize_t>& shape) {

    auto zone = std::make_shared<Zone>(name);
    auto zoneNode = std::const_pointer_cast<Node>(zone->selfPtr());
    if (!zoneNode) {
        throw py::value_error("makeCartesianZone: expected heap-managed zone");
    }

    auto coordinates = std::make_shared<Node>("GridCoordinates", "GridCoordinates_t");
    coordinates->attachTo(zoneNode);

    auto addCoordinate = [&](const std::string& coordName) {
        py::array_t<double> values(shape);
        auto coord = std::make_shared<Node>(coordName, "DataArray_t");
        coord->setData(std::make_shared<Array>(arraybridge::arrayFromPyArray(values)));
        coord->attachTo(coordinates);
    };

    addCoordinate("CoordinateX");
    addCoordinate("CoordinateY");
    addCoordinate("CoordinateZ");

    zone->updateShape();
    return zone;
}

} // namespace

void test_base_new_base_dimensions() {
    auto zoneA = makeCartesianZone("ZoneA", {3, 3, 3});
    auto zoneB = makeCartesianZone("ZoneB", {4, 3, 2});

    auto base = newBase("Base", {zoneA, zoneB});
    if (base->name() != "Base") throw py::value_error("expected base name Base");
    if (base->type() != "CGNSBase_t") throw py::value_error("expected CGNSBase_t");
    if (base->dim() != 3) throw py::value_error("expected cell dimension 3");
    if (base->physicalDimension() != 3) throw py::value_error("expected physical dimension 3");
    if (base->numberOfZones() != 2) throw py::value_error("expected two zones");
    if (zoneA->parent().lock().get() != base.get()) throw py::value_error("zone should be attached to base");
}

void test_base_aggregates_zones() {
    auto zoneA = makeCartesianZone("ZoneA", {3, 3, 3});
    auto zoneB = makeCartesianZone("ZoneB", {4, 3, 2});
    auto base = newBase("Base", {zoneA, zoneB});

    if (!base->isStructured()) throw py::value_error("base should be structured");
    if (base->numberOfPoints() != zoneA->numberOfPoints() + zoneB->numberOfPoints()) {
        throw py::value_error("base point count mismatch");
    }
    if (base->numberOfCells() != zoneA->numberOfCells() + zoneB->numberOfCells()) {
        throw py::value_error("base cell count mismatch");
    }

    auto fields = base->newFields({{"q", 7.0}});
    if (fields.size() != 2) throw py::value_error("expected field output for two zones");
    if (zoneA->field("q")->size() != zoneA->numberOfPoints()) {
        throw py::value_error("zoneA field size mismatch");
    }
}

void test_tree_aggregates_bases_and_zones() {
    auto zoneA = makeCartesianZone("ZoneA", {3, 3, 3});
    auto zoneB = makeCartesianZone("ZoneB", {4, 3, 2});
    auto base = newBase("Base", {zoneA});

    auto tree = std::make_shared<Tree>();
    tree->add(base);
    tree->add(zoneB);

    if (tree->type() != "CGNSTree_t") throw py::value_error("expected CGNSTree_t");
    if (tree->numberOfBases() != 1) throw py::value_error("expected one base");
    if (tree->numberOfZones() != 2) throw py::value_error("expected two zones");
    if (tree->numberOfPoints() != zoneA->numberOfPoints() + zoneB->numberOfPoints()) {
        throw py::value_error("tree point count mismatch");
    }

    auto names = tree->zoneNames();
    if (names.size() != 2 || names[0] != "ZoneA" || names[1] != "ZoneB") {
        throw py::value_error("unexpected tree zone names");
    }
}

void test_new_base_rejects_incoherent_dimensions() {
    auto zone3D = makeCartesianZone("Zone3D", {3, 3, 3});
    auto zone2D = makeCartesianZone("Zone2D", {3, 3});

    bool raised = false;
    try {
        (void)newBase("BadBase", {zone3D, zone2D});
    } catch (const std::invalid_argument&) {
        raised = true;
    }

    if (!raised) throw py::value_error("newBase should reject mixed cell dimensions");
}

void test_newZoneFromArrays() {
    py::array_t<double> x({3, 2});
    py::array_t<double> y({3, 2});
    py::array_t<double> z({3, 2});
    py::array_t<double> pressure({3, 2});

    auto zone = newZoneFromArrays(
        "Block",
        {"x", "y", "z", "Pressure"},
        {
            std::make_shared<Array>(arraybridge::arrayFromPyArray(x)),
            std::make_shared<Array>(arraybridge::arrayFromPyArray(y)),
            std::make_shared<Array>(arraybridge::arrayFromPyArray(z)),
            std::make_shared<Array>(arraybridge::arrayFromPyArray(pressure))
        });

    if (zone->name() != "Block") throw py::value_error("expected zone name Block");
    if (zone->type() != "Zone_t") throw py::value_error("expected Zone_t");
    if (zone->dim() != 2) throw py::value_error("expected 2D zone");
    if (zone->numberOfPoints() != 6) throw py::value_error("expected six points");
    if (zone->numberOfCells() != 2) throw py::value_error("expected two cells");

    auto coordinates = zone->pick().childByName("GridCoordinates");
    if (!coordinates) throw py::value_error("expected GridCoordinates");
    if (!coordinates->pick().childByName("CoordinateX")) throw py::value_error("expected CoordinateX");
    if (!coordinates->pick().childByName("CoordinateY")) throw py::value_error("expected CoordinateY");
    if (!coordinates->pick().childByName("CoordinateZ")) throw py::value_error("expected CoordinateZ");

    auto fields = zone->pick().childByName("FlowSolution");
    if (!fields) throw py::value_error("expected FlowSolution");
    auto pressureNode = fields->pick().childByName("Pressure");
    if (!pressureNode) throw py::value_error("expected Pressure field");
    if (pressureNode->dataPtr()->shape() != std::vector<size_t>({3, 2})) {
        throw py::value_error("unexpected Pressure shape");
    }
}

File: tests/python/cgns/test_base_tree.py
import numpy as np
import pytest

from noder import read
from noder.core import Node, nodeToPyCGNS, pyCGNSToNode
from noder.cgns import (Base, Tree, Zone, add, merge, new_tree, new_base,
                        new_zone_from_arrays, new_zone_from_dict)

@pytest.fixture
def pycgns_list():
    pylist = [
        'CGNSTree',
        None,
        [

            ['Base1',
            np.array([3,3],dtype=np.int32,order='F'),
            [
                ['ZoneA',
                np.array([[16,9,0]],dtype=np.int32,order='F'),
                [
                    ['GridCoordinates',
                    None,
                    [
                        ['CoordinateX',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateY',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateZ',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                    ],
                    'GridCoordinates_t'
                    ]
                ],
                'Zone_t'],
                ['ZoneB',
                np.array([[16,9,0]],dtype=np.int32,order='F'),
                [
                    ['GridCoordinates',
                    None,
                    [
                        ['CoordinateX',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateY',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateZ',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                    ],
                    'GridCoordinates_t'
                    ]
                ],
                'Zone_t'],
            ],
            'CGNSBase_t'],

            ['Base2',
            np.array([3,3],dtype=np.int32,order='F'),
            [
                ['ZoneD',
                np.array([[16,9,0]],dtype=np.int32,order='F'),
                [
                    ['GridCoordinates',
                    None,
                    [
                        ['CoordinateX',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateY',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateZ',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                    ],
                    'GridCoordinates_t'
                    ]
                ],
                'Zone_t'],
                ['ZoneE',
                np.array([[16,9,0]],dtype=np.int32,order='F'),
                [
                    ['GridCoordinates',
                    None,
                    [
                        ['CoordinateX',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateY',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                        ['CoordinateZ',
                            np.zeros((4,4),dtype=float, order='F'),
                            [],
                            'DataArray_t'
                        ],
                    ],
                    'GridCoordinates_t'
                    ]
                ],
                'Zone_t'],
            ],
            'CGNSBase_t'],


        ],
        'CGNSTree_t'
    ]

    return pylist


def make_cart(name="block", shape=(3, 3, 3)):
    values = np.ones(shape, dtype=np.float64)

    zone = Zone(name)
    coordinates = Node("GridCoordinates", "GridCoordinates_t")

    for coordinate in ["CoordinateX", "CoordinateY", "CoordinateZ"]:
        child = Node(coordinate, "DataArray_t")
        child.set_data(values.copy())
        coordinates.add_child(child)

    zone.add_child(coordinates)
    zone.update_shape()
    return zone


def test_base_init_and_new_base():
    zone_a = make_cart("ZoneA")
    zone_b = make_cart("ZoneB", (4, 3, 2))

    base = new_base("Base", [zone_a, zone_b])
    assert isinstance(base, Base)
    assert base.name() == "Base"
    assert base.type() == "CGNSBase_t"
    assert base.dim() == 3
    assert base.physical_dimension() == 3
    assert base.number_of_zones() == 2
    assert [z.name() for z in base.zones()] == ["ZoneA", "ZoneB"]


def test_cgns_helpers_are_exported_from_package_root():
    from noder import add as root_add
    from noder import Base as RootBase
    from noder import merge as root_merge
    from noder import Tree as RootTree
    from noder import Zone as RootZone
    from noder import new_base as root_new_base
    from noder import new_zone_from_arrays as root_new_zone_from_arrays
    from noder import new_zone_from_dict as root_new_zone_from_dict

    assert RootBase is Base
    assert RootTree is Tree
    assert RootZone is Zone
    assert root_new_base is new_base
    assert root_new_zone_from_arrays is new_zone_from_arrays
    assert root_new_zone_from_dict is new_zone_from_dict
    assert root_add is add
    assert root_merge is merge


def test_new_zone_from_arrays_and_dict():
    x, y = np.meshgrid(np.arange(3.0), np.arange(2.0), indexing="ij")
    z = np.zeros_like(x)
    pressure = x + y

    zone = new_zone_from_arrays("Block", ["x", "y", "z", "Pressure"], [x, y, z, pressure])
    assert isinstance(zone, Zone)
    assert zone.name() == "Block"
    assert zone.dim() == 2
    assert zone.number_of_points() == 6
    assert zone.number_of_cells() == 2
    assert zone.x().shape == (3, 2)
    assert np.array_equal(zone.field("Pressure", behavior_if_not_found="raise"), pressure)

    zone_from_dict = new_zone_from_dict("DictBlock", {"x": x, "y": y, "z": z, "Pressure": pressure})
    assert isinstance(zone_from_dict, Zone)
    assert zone_from_dict.name() == "DictBlock"
    assert np.array_equal(zone_from_dict.field("Pressure", behavior_if_not_found="raise"), pressure)


def test_add_and_merge_helpers():
    zone_a = make_cart("A")
    zone_b = make_cart("B")
    zone_c = make_cart("C")

    tree = add(zone_a, SecondBase=[zone_b, zone_c])
    assert isinstance(tree, Tree)
    assert tree.number_of_bases() == 2
    assert tree.base_names() == ["Base", "SecondBase"]
    assert [zone.name() for zone in tree.base("Base").zones()] == ["A"]
    assert [zone.name() for zone in tree.base("SecondBase").zones()] == ["B", "C"]

    left = Node("Root")
    left.add_child(Node("left"))
    right = Node("Root")
    right.add_child(Node("right"))

    merged = merge([left, right])
    assert merged is not left
    assert merged.name() == "Root"
    assert set(merged.get_children_names()) == {"left", "right"}


def test_new_base_rejects_incoherent_zone_dimensions():
    with pytest.raises(ValueError):
        new_base("Base", [make_cart("Zone3D"), make_cart("Zone2D", (3, 3))])


def test_base_aggregates_fields_and_counts():
    base = new_base("Base", [make_cart("ZoneA"), make_cart("ZoneB", (4, 3, 2))])

    assert base.is_structured()
    assert base.number_of_points() == 27 + 24
    assert base.number_of_cells() == 8 + 6

    fields = base.new_fields("q")
    assert set(fields) == {"ZoneA", "ZoneB"}
    assert fields["ZoneA"]["q"].shape == (3, 3, 3)
    assert fields["ZoneB"]["q"].shape == (4, 3, 2)

    all_fields = base.all_fields(include_coordinates=False)
    assert set(all_fields) == {"ZoneA", "ZoneB"}
    assert set(all_fields["ZoneA"]) == {"q"}


def test_tree_aggregates_bases_and_zones():
    base = new_base("Base", [make_cart("ZoneA")])
    zone_b = make_cart("ZoneB", (4, 3, 2))

    tree = Tree()
    tree.add(base)
    tree.add(zone_b)

    assert tree.name() == "CGNSTree"
    assert tree.type() == "CGNSTree_t"
    assert tree.number_of_bases() == 1
    assert tree.number_of_zones() == 2
    assert tree.zone_names() == ["ZoneA", "ZoneB"]
    assert tree.number_of_points() == 27 + 24

    fields = tree.new_fields(["q", "r"])
    assert set(fields) == {"ZoneA", "ZoneB"}
    assert set(fields["ZoneA"]) == {"q", "r"}


def test_pycgns_converter_returns_specialized_base_and_tree():
    tree = Tree()
    tree.add(new_base("Base", [make_cart("ZoneA")]))

    pycgns_tree = nodeToPyCGNS(tree)
    tree_again = pyCGNSToNode(pycgns_tree)

    assert isinstance(tree_again, Tree)
    assert isinstance(tree_again.children()[0], Base)
    assert isinstance(tree_again.children()[0].children()[0], Zone)


def test_pycgns_converter_with_custom_tree(pycgns_list):

    t = pyCGNSToNode(pycgns_list)
    reached = False
    assert isinstance(t,Tree)
    for b in t.bases():
        assert isinstance(b,Base)
        for z in b.zones():
            assert isinstance(z,Zone)
            for n in z.pick().all_by_and(type='DataArray_t'):
                assert isinstance(n, Node)
                reached = True
    assert reached


def test_pycgns_converter_with_hdf5_root_label(pycgns_list):
    pycgns_list[0] = "HDF5 MotherNode"
    pycgns_list[3] = "Root Node of HDF5 File"

    t = pyCGNSToNode(pycgns_list)

    assert isinstance(t, Tree)
    assert t.name() == "HDF5 MotherNode"
    assert t.type() == "Root Node of HDF5 File"
    assert all(isinstance(base, Base) for base in t.bases())
    assert all(isinstance(zone, Zone) for base in t.bases() for zone in base.zones())

def test_types_after_read(tmp_path, pycgns_list):
    t = pyCGNSToNode(pycgns_list)
    filename = str(tmp_path/"test.cgns")
    t.write(filename)

    tr = read(filename)
    reached = False
    assert isinstance(tr,Tree), "expected a Tree"
    for b in tr.bases():
        assert isinstance(b,Base), "expected a Base"
        for z in b.zones():
            assert isinstance(z,Zone), "expected a Zone"
            for n in z.pick().all_by_and(type='DataArray_t'):
                assert isinstance(n, Node)
                reached = True
    assert reached

def test_new_tree():

    zeros3x1 = np.zeros(3,dtype=float,order='F')
    zeros3x3 = np.zeros((3,3),dtype=float,order='F')
    zeros3x3x3 = np.zeros((3,3,3),dtype=float,order='F')
    zone1 = new_zone_from_arrays('zone1',['x','y','z'],[zeros3x3,zeros3x3,zeros3x3])
    zone2 = new_zone_from_arrays('zone2',['x','y','z'],[zeros3x3,zeros3x3,zeros3x3])
    zone3 = new_zone_from_arrays('zone3',['x','y','z'],[zeros3x1,zeros3x1,zeros3x1])
    zone4 = new_zone_from_arrays('zone4',['x','y','z'],[zeros3x1,zeros3x1,zeros3x1])
    zone5 = new_zone_from_arrays('zone5',['x','y','z'],[zeros3x3x3,zeros3x3x3,zeros3x3x3])

    t = new_tree(BaseA=[zone1,zone2], BaseB=[zone3,zone4],BaseC=zone5)

    assert isinstance(t, Tree)
    bases = t.bases()
    assert len(bases)==3

    BaseA = bases[0]
    assert isinstance(BaseA,Base)
    # TODO check if BaseA value accounts for dim=2
    zones_of_BaseA = BaseA.zones()
    assert len(zones_of_BaseA) == 2
    zone1_picked = zones_of_BaseA[0]
    assert isinstance(zone1_picked,Zone)
    assert zone1_picked is zone1
    zone2_picked = zones_of_BaseA[1]
    assert isinstance(zone2_picked,Zone)
    assert zone2_picked is zone2

    BaseB = bases[1]
    assert isinstance(BaseB,Base)
    # TODO check if BaseB value accounts for dim=1
    zones_of_BaseB = BaseB.zones()
    assert len(zones_of_BaseB) == 2
    zone3_picked = zones_of_BaseB[0]
    assert isinstance(zone3_picked,Zone)
    assert zone3_picked is zone3
    zone4_picked = zones_of_BaseB[1]
    assert isinstance(zone4_picked,Zone)
    assert zone4_picked is zone4

    BaseC = bases[2]
    assert isinstance(BaseC,Base)
    # TODO check if BaseC value accounts for dim=3
    zones_of_BaseC = BaseC.zones()
    assert len(zones_of_BaseC) == 1
    zone5_picked = zones_of_BaseC[0]
    assert isinstance(zone5_picked,Zone)
    assert zone5_picked is zone5
